File: frontend/src/api/banking.ts
// Enable Banking HTTP helpers. Kept alongside the other ``api/`` modules so
// pages can import from one place.

import { api } from "./client";
import type { Scope } from "./types";

export type Aspsp = {
  name: string;
  country: string;
  logo: string | null;
};

export type BankConnection = {
  id: number;
  aspsp_name: string;
  aspsp_country: string;
  valid_until: string | null;
  last_synced_at: string | null;
  last_error_code: string | null;
};

export type ConnectResponse = {
  authorization_url: string;
  state: string;
};

export type SkippedAccount = {
  name: string;
  reason: string;
};

export type ConnectionSyncResult = {
  connection_id: number;
  aspsp_name: string;
  added: number;
  modified: number;
  skipped_pending: number;
  skipped_non_eur: number;
  skipped_unknown_account: number;
  skipped_unparseable_date: number;
  error_code: string | null;
  touched_account_ids: number[];
};

export type SyncStatus = {
  mode: "manual" | "auto";
  max_per_day: number;
  used_today: number;
  remaining_today: number;
  last_run_at: string | null;
  next_auto_sync_at: string | null;
};

export type GlobalSyncResponse = {
  status: "ok" | "partial" | "error";
  connections: ConnectionSyncResult[];
  quota: SyncStatus;
};

export type CallbackResponse = {
  connection: BankConnection;
  account_ids: number[];
  skipped_accounts: SkippedAccount[];
  sync: ConnectionSyncResult;
};

export const bankingApi = {
  listAspsps: () => api<Aspsp[]>("/api/banking/aspsps"),
  connect: (aspsp_name: string, aspsp_country: string, scope: Scope = "personal") =>
    api<ConnectResponse>("/api/banking/connections", {
      method: "POST",
      body: { aspsp_name, aspsp_country, scope },
    }),
  completeCallback: (code: string, state: string) =>
    api<CallbackResponse>("/api/banking/connections/callback", {
      method: "POST",
      body: { code, state },
    }),
  syncNow: () => api<GlobalSyncResponse>("/api/banking/sync", { method: "POST" }),
  syncStatus: () => api<SyncStatus>("/api/banking/sync/status"),
  listConnections: () => api<BankConnection[]>("/api/banking/connections"),
  unlinkConnection: (connectionId: number) =>
    api<void>(`/api/banking/connections/${connectionId}`, { method: "DELETE" }),
};

File: backend/app/schemas/banking.py
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field


class AspspResponse(BaseModel):
    name: str
    country: str
    logo: str | None = None


class ConnectRequest(BaseModel):
    aspsp_name: str = Field(min_length=1, max_length=255)
    aspsp_country: str = Field(min_length=2, max_length=2)
    scope_id: int


class ConnectResponse(BaseModel):
    authorization_url: str
    state: str


class CallbackRequest(BaseModel):
    code: str = Field(min_length=1)
    state: str = Field(min_length=1)


class SkippedAccount(BaseModel):
    name: str
    reason: str


class BankConnectionResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    aspsp_name: str
    aspsp_country: str
    valid_until: datetime | None
    last_synced_at: datetime | None
    last_error_code: str | None


class ConnectionSyncResult(BaseModel):
    connection_id: int
    aspsp_name: str
    added: int
    modified: int
    skipped_pending: int
    skipped_non_eur: int
    skipped_unknown_account: int
    skipped_unparseable_date: int
    skipped_deleted: int
    error_code: str | None = None
    touched_account_ids: list[int]


class SyncStatusResponse(BaseModel):
    mode: str
    max_per_day: int
    used_today: int
    remaining_today: int
    last_run_at: datetime | None
    # Only set in auto mode: when the scheduler will next consider a run.
    next_auto_sync_at: datetime | None


class GlobalSyncResponse(BaseModel):
    status: str
    connections: list[ConnectionSyncResult]
    quota: SyncStatusResponse


class CallbackResponse(BaseModel):
    connection: BankConnectionResponse
    account_ids: list[int]
    skipped_accounts: list[SkippedAccount]
    sync: ConnectionSyncResult
